docs/styles.rst

```diff
@@ -0,0 +1,15 @@
+Styles
+======
+
+VSG supports several predefined styles.
+They can be accessed by using the **--style** command line option.
+
++---------------+--------------------------------------+
+| Configuration | Description                          |
++---------------+--------------------------------------+
+| legacy        | Rules based on the pre 2.0.0 release |
+| minimal       | Only indenting is checked            |
```

Comment:
So before this feature I would have resisted defaulting to a minimum style because creating configurations can be time consuming.  However, with these styles it makes it easier to users to pick a style that closely matches their own, or even pick one from a list of styles.

If we switch to a minimal style we could add a note to the end of a run that states something like:
```mono
NOTE:  Consider starting with a built in style using the --style argument.
   legacy    : rules based on the pre 2.0.0 release
   lowercase : vhdl keywords, signal names etc... are lowercase
```
... if `--style` or `--configuration` was not used.  Something to nudge the user in the right direction.

So instead of VSG providing a single opinion on what your code should look like and forcing the user to modify that opinion, VSG would provide no opinion and allow the user to pick one from those provided.  That selection is a starting point that can be further modified with the `--configuration` option.

Reply:
The problem with the defaults, as we have seen, is that it is hard to maintain backward compatibility in some cases. The default style (no `--style` or `--configuration`) probably shouldn't be allowed. Maybe we should immediately return and print info to the user that he should use `--style`, `--configuration` or both parameters? Or maybe `style` should become positional argument?

Reply:
hmmm....making `--style` a required argument is an interesting idea.  It would force the user to investigate what current styles are available.  Probably best not do define a default value either.

Reply:
On the other hand, someone might to define his own style, then forcing him to use --style parameter does not seem like a good solution.

Reply:
Maybe there is a `none` style and that is the default?

If I was running this for the first time, I would want VSG to do something or least tell me what I need to do to get it to do something.

Reply:
We might enforce the user to provide one of:
1. `--style`
2. `--configuration`
3. `--style` and `--configuration`

If user provides neither `--style` nor `--configuration` he will get an error.

Reply:
If we keep `--style` optional, then the default behavior will be however the rules are currently defined.   I think that is a reasonable position to take.  It addresses my "VSG should do something" issue.

Then that raises a question of why even have this feature.  I could always just use the `--configuration` option.

Reply:
Ok, I have one idea. Style becomes positional argument. This way we enforce user to do something to get to know about styles. We provide one special style named `minimal` or `base` with indentation only.

There is still one thing in my mind. What if user wants to define his own style?
We can tell him to use minimal/base style and apply configuration. But it requires him to call vsg in following way `vsg minimal/base --configuraiton my_config`, quite long, on the other hand he can always define an alias.
We can also accept PR with styles. I do not think there will be many.

Reply:
k, I think I am going to go back on something I said before:
> So instead of VSG providing a single opinion on what your code should look like and forcing the user to modify that opinion, VSG would provide no opinion and allow the user to pick one from those provided.

VSG should have an opinion on what makes well formatted VHDL code.  That opinion is it's base rule set out of the box.

There are over 350 rules and for the legacy style I modified 26 of them (I will probably need to modify more).  They were all capitalization and alignment rules.  Those 26 represent just 7.4% of the total number of rules.  I want to believe that means the majority of the rules would be agreeable to most people.

The `--style` is there to provide the user with other possibilities.

It is almost like VSG should say...

"I believe this is the most readable formatted VHDL code.  If you do not agree, then I have a couple of other styles that people like.  If you do not like any of those, then you can make make up your own style."

Reply:
Do I understand correctly, the legacy style will be the default one?

Reply:
The current rule base, with the lower casing, would be the default.  Plus anything new that gets added, like #393, #378, etc...  As those new rule sets get implemented, then the other styles would be updated.

Reply:
I have one more idea. Lets make the legacy style the default. We disable all the rules in the code (no rule is enabled in any .py file). When `--style` is not passed to the vsg, then vsg chooses the legacy style as the default one. What do you think?

Reply:
This might be the point at which we need to test out what I have implemented.  I just pushed a a version to this branch with three styles:  base, disabled and legacy.  I set the base to be the default.

I ran a little experiment on the `spi_slave.vhd` file and switched back and forth between the base and legacy styles.  It is pretty cool to be able to switch between them quickly.

How about you checkout this branch and give it try.  It might help solidify an approach.

Reply:
Ok, but tomorrow. I got to go right now.

Reply:
`python setup.py install --user` does not install styles, so vsg does not work.
```
FileNotFoundError: [Errno 2] No such file or directory: '/home/mkru/.local/lib/python3.8/site-packages/vsg-1.2.0-py3.8.egg/vsg/styles'
```

I think `styles` directory should be placed next to `vsg` directory, not inside. `vsg` should contain Python sources related to vsg.

Reply:
> `python setup.py install --user` does not install styles, so vsg does not work.

That is frustrating.  It must have something to do with the `find_packages` function in the `setup.py` file.

> I think styles directory should be placed next to vsg directory, not inside. vsg should contain Python sources related to vsg.

Will setup.py include them in the install?

Reply:
We have to investigate this, but I have seen projects, where directories sitting next to the directory with actual Python code are installed, so people somehow achieve this.

Reply:
@m-kru I figured out the issue.  I did not include an `__init__.py` file in the `vsg/styles` directory.  The function `find_packages` will only include directories with that file.

Reply:
I now copies styles directory, however without styles .yml files.

Reply:
I think I got it this time.  Please try again.

Reply:
Ok, works.
So final questions.
1. What we do in case when user does not provide `--style` argument, do we use some style as the default one, or do we inform him, that he must provide `--style` argument? If we use some default style, then I propose the legacy style to be the default one. We can even name this style default.
2.  Do we provide any style with minimal number of rules enabled? 

In my humble opinion disabled style makes no sense.

Btw. I thnk the note at the end misses `the` word.
```
NOTE: Refer to THE online documentation at https://vhdl-style-guide.readthedocs.io/en/latest/index.html for more information.
```

Reply:
> What we do in case when user does not provide `--sytle` argument...
We would default to a style, which is currently `base`.
> If we use some default style, then I propose the legacy style...
The legacy style has all of my capitalization and alignment formatting.  That is just an artifact of where I learned to write VHDL.  I think VSG should focus on what provides the most readable VHDL, which I think is currently the base rules.

I should rename `legacy` to something else, but I am having trouble thinking up a name.

> Do we provide any style with minimal number of rules enabled?

I think we should try.  I am a little worried about Phase 1 though:
![image](https://user-images.githubusercontent.com/24329243/82236038-2566e380-98f9-11ea-8681-42a7f7a803e2.png)

Phase 1 we could disable certain rules, maybe all of them?
Phase 2 we could skip (disable all rules in phase 2)
Phase 3 we could skip (disable all rules in phase 3)
Phase 4 we would do
Phases 5 through 7 we would skip (disable all rules in those phases)

I currently do not have a method to skip phases.  Setting up a configuration to disable the rules in the phase would require constant maintenance.  I could put a skip phases or execute phases parameter in the style YAML file:

```yaml
skip_phases:
    - 1
    - 2
    - 3
    - 5
    - 6
    - 7
```
Pass that list to the `rule_list` object and let it go.

Reply:
I guess some rules from phases 2-7 may fail, if the structure is weird. I would opt for enabling phase 1 and 4 in the minimal/base style. But I do not know if all rules should be enabled I would have to read them first.

You still have not answered if `--style` will be mandatory?

Reply:
> You still have not answered if --style will be mandatory?

I don't think it should be mandatory.  We should default to the base rule set and you can override it with another style.

Reply:
well, I implemented the `skip_phase` and ran just phase 4 (Indent) on the PIC.vhd file and it seemed to work.  The file is well structured though.  I will try other files.

Reply:
I ran against every file in the code_examples and it actually did fairly well.  There are a couple of issues.
  1)  There are some alignment rules that should probably be ran in the indent phase:
          process_020
          concurrent_003
  2)  Ending parenthesis in generic and port statements when they are not on their own line have indent issues.

Other than that it looked pretty good.  I will work on committing the changes.

Reply:
1. Trying to run alignment rules in indent phase probably means that something is not implemented in proper way.
2. This is somehow related to the 1.

I think we should think for  a while on how we handle indentation and how we handle alignment in case of multi line issues, and what is what. Right now I wonder how would I handle indentation/alignment in the signal declaration approach I would like to apply. https://github.com/jeremiah-c-leary/vhdl-style-guide/issues/290

Reply:
> I think we should think for a while on how we handle indentation and how we handle alignment in case of multi line issues, and what is what.

That is a good point.

```mono
I      a <= b or
A         c and d when
A         e = 1 else
A         a;
```
Where I = Indent and A = alignment.
```mono
I  CLK_PROC : process (a ,b
A                      c, d) is
I   begin
I   end process CLK_PROC;
```

```mono
I   case state
I     when idle =>
I       n_state <= write;
I       n_write_en <= a or b or
A                     c and d;
I    when write =>
```

It would seem that it a line starts a VHDL element? that it would be an indent.  If a line continues an element then it would be an alignment.  This distinction only applies to the beginning of the line.

```mono
I  entity FIFO is
I  port (
I   I_WR_EN : in std_logic;
I   I_RD_EN : in std_logic
I  );
I  end entity FIFO;
```

Reply:
@m-kru I just pushed the minimal style and tests.  You can see some examples of alignments not occuring in some of the files, due to the alignment phase being skipped.

Reply:
Hmm, I like your idea about marking lines with I and A. However right now it does not happen when the file is parsed, does it?

Reply:
The lines are not marked `I` and `A` specifically, but it can be inferred from the `line.indentLevel` attribute:

```mono
2     case state
3       when idle =>
4         n_state <= write;
4         n_write_en <= a or b or
None                    c and d;
3       when write =>
```

Reply:
> Right now I wonder how would I handle indentation/alignment in the signal declaration approach I would like to apply. #290

In this case the indent/alignment would be:

```mono
2      signal write_en,
None          read_en,
None          full,
None          almost_full : std_logic;
```

Reply:
hey @m-kru just wanted to ping you about this PR.  Do you think this is done enough to merge into master?

Reply:
I will analyze it again today.